fuel-p2p/src/request_response/codec.rs

```diff
@@ -0,0 +1,110 @@
+use super::messages::{RequestMessage, ResponseMessage};
+use async_trait::async_trait;
+use futures::{AsyncRead, AsyncWriteExt};
+use libp2p::{
+    core::{
+        upgrade::{read_length_prefixed, write_length_prefixed},
+        ProtocolName,
+    },
+    request_response::RequestResponseCodec,
+};
+use std::io;
+
+pub const REQUEST_RESPONSE_PROTOCOL_ID: &[u8] = b"/fuel/req_res/0.0.1";
+/// Max Size in Bytes of the messages
+// todo: this will be defined once Request & Response Messages are clearly defined
+const MAX_REQUEST_SIZE: usize = std::mem::size_of::<RequestMessage>();
+const MAX_RESPONSE_SIZE: usize = std::mem::size_of::<ResponseMessage>();
+
+#[derive(Debug, Clone)]
+pub struct MessageExchangeProtocol;
+
+impl ProtocolName for MessageExchangeProtocol {
+    fn protocol_name(&self) -> &[u8] {
+        REQUEST_RESPONSE_PROTOCOL_ID
+    }
+}
+
+#[derive(Debug, Clone)]
+pub struct MessageExchangeCodec {}
+
+// Since Bincode does not support async reads or writes out of the box
+// We prefix Request & Response Messages with the length of the data in bytes
```

Comment:
I was confused by the `stream.close()`, but realized from the examples that the socket is a dedicated substream for this request-response flow. Copied this description from their docs since I figured it might help with readability here: 

```suggestion
// We prefix Request & Response Messages with the length of the data in bytes
/// We expect the substream to be properly closed when response channel is dropped.
/// Since the request protocol used here expects a response, the sender considers this
/// early close as a protocol violation which results in the connection being closed.
/// If the substream were not properly closed when dropped, the sender would instead
/// run into a timeout waiting for the response.
```

Reply:
yes, it's a one off